[Messaging] Add IBMMQ-AMQ example

avano opened the pull request.

avano commented:
@zhfeng can you review pls?
@aldettinger fyi

zhfeng commented on message-bridge/eclipse-formatter-config.xml:
Why we need this file? I think it is related to Eclipse IDE?

avano replied:
it is used in the formatter plugin [here](https://github.com/avano/camel-quarkus-examples/blob/ibmmq-amq/message-bridge/pom.xml#L140). it seems to be present in every example, so I included it here as well

zhfeng replied:
OK - sorry, I have not noticed it.

zhfeng commented on message-bridge/src/main/java/org/acme/message/bridge/MessageBridgeRoute.java:
```diff
@@ -0,0 +1,46 @@
+                .to("amq:queue:{{amq.queue}}")
+                .process(ex -> {
+                    if (ex.getIn().getBody(String.class).toLowerCase().contains("rollback")) {
+                        if (!ex.getIn().getHeader("JMSRedelivered", Boolean.class)) {
```
Why we check `JMSRedelivered` here?

aldettinger replied:
The idea might be to rollback the transaction only once. So that we see a rollback followed by successful commit. Maybe a comment could help readability.

avano replied:
yes, @aldettinger is correct, just some way how to proceed with the transaction after the rollback, I'll add a comment there

zhfeng commented on message-bridge/src/main/resources/application.properties:
```diff
@@ -0,0 +1,34 @@
+
+quarkus.pooled-jms.transaction=xa
+quarkus.pooled-jms.pooling.enabled=true
+quarkus.pooled-jms.max-connections=8
```
I'd like to add `quarkus.transaction-manager.enable-recovery=true` and it could be helpful to test the crash crash secenario.

avano replied:
thanks, will check that out as well

zhfeng commented:
Thanke @avano and I just add my comments. The other thing is that license header is missing in some files. Please check them!

jamesnetherton commented on message-bridge/pom.xml:
```diff
@@ -0,0 +1,274 @@
+        <!-- REST endpoint -->
+        <dependency>
+            <groupId>org.apache.camel.quarkus</groupId>
+            <artifactId>camel-quarkus-servlet</artifactId>
```
You can remove this as `camel-quarkus-rest` includes `camel-quarkus-platform-http`. 

You can remove the servlet config from `application.properties` afterwards.

jamesnetherton commented:
You'll need to add a new entry into `docs/modules/ROOT/attachments/examples.json`.

aldettinger commented on message-bridge/README.adoc:
```diff
@@ -0,0 +1,107 @@
+  -e LICENSE=accept \
+  -e MQ_QMGR_NAME=QM1 \
+  -e MQ_APP_PASSWORD=passw0rd \
+  -p 1414:1414
```
Is there a missing "\" at the end of this line ?

avano replied:
thanks, you're right

aldettinger commented:
When executing `mvn clean compile quarkus:dev`, There is a WARN log:

```
================================================================================
CDI: programmatic lookup problem detected
-----------------------------------------
At least one bean matched the required type and qualifiers but was marked as unused and removed during build

Stack frame: io.quarkus.arc.runtime.BeanContainerImpl.beanInstanceFactory(BeanContainerImpl.java:26)
Required type: class org.apache.camel.component.servlet.CamelHttpTransportServlet
Required qualifiers: [@jakarta.enterprise.inject.Default()]
Removed beans:
	- CLASS bean  [types=[class org.apache.camel.http.common.CamelServlet, class jakarta.servlet.http.HttpServlet, class jakarta.servlet.GenericServlet, class org.apache.camel.component.servlet.CamelHttpTransportServlet, interface jakarta.servlet.Servlet, interface java.io.Serializable, interface org.apache.camel.http.common.HttpRegistryProvider, interface jakarta.servlet.ServletConfig], qualifiers=null]
Solutions:
	- Application developers can eliminate false positives via the @Unremovable annotation
	- Extensions can eliminate false positives via build items, e.g. using the UnremovableBeanBuildItem
	- See also https://quarkus.io/guides/cdi-reference#remove_unused_beans
	- Enable the DEBUG log level to see the full stack trace
```

aldettinger commented on message-bridge/pom.xml:
```diff
@@ -0,0 +1,274 @@
+                    <version>${maven-surefire-plugin.version}</version>
+                    <configuration>
+                        <failIfNoTests>false</failIfNoTests>
+                        <systemProperties>
```
There is this warning:

```
[WARNING]  Parameter 'systemProperties' is deprecated: Use systemPropertyVariables instead.
```

This one mayb be solved like this:
https://github.com/apache/camel-quarkus-examples/commit/589b7e3470ede35f2224b56eb3f5918bc00c8b87

aldettinger commented on message-bridge/src/test/java/org/acme/message/bridge/MessageBridgeTest.java:
```diff
@@ -0,0 +1,83 @@
+
+@QuarkusTest
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class MessageBridgeTest extends CamelQuarkusTestSupport {
```
Really awesome to have tests :+1: 

Would it be even better to transform those tests ala integration test in another pr ? Maybe it would offer more coverage to detect regressions + it could also help on the path to test in native mode. Let's discuss.

avano replied:
sure, I can rework in this PR as well, since there are changes needed anyways

avano commented:
> When executing `mvn clean compile quarkus:dev`, There is a WARN log:
> 
> ```
> ================================================================================
> CDI: programmatic lookup problem detected
> -----------------------------------------
> At least one bean matched the required type and qualifiers but was marked as unused and removed during build
> 
> Stack frame: io.quarkus.arc.runtime.BeanContainerImpl.beanInstanceFactory(BeanContainerImpl.java:26)
> Required type: class org.apache.camel.component.servlet.CamelHttpTransportServlet
> Required qualifiers: [@jakarta.enterprise.inject.Default()]
> Removed beans:
> 	- CLASS bean  [types=[class org.apache.camel.http.common.CamelServlet, class jakarta.servlet.http.HttpServlet, class jakarta.servlet.GenericServlet, class org.apache.camel.component.servlet.CamelHttpTransportServlet, interface jakarta.servlet.Servlet, interface java.io.Serializable, interface org.apache.camel.http.common.HttpRegistryProvider, interface jakarta.servlet.ServletConfig], qualifiers=null]
> Solutions:
> 	- Application developers can eliminate false positives via the @Unremovable annotation
> 	- Extensions can eliminate false positives via build items, e.g. using the UnremovableBeanBuildItem
> 	- See also https://quarkus.io/guides/cdi-reference#remove_unused_beans
> 	- Enable the DEBUG log level to see the full stack trace
> ```

will be fixed by removing the servlet dependency, as mentioned by James

aldettinger commented:
Executing locally, there are some files that should probably be git ignored:

```
[main_upstream @ message-bridge]$ git status
HEAD detached at avano/ibmmq-amq
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	FFDC/
	mqjms.log.0
```

The FFDC folder looks like a kind of internal logs to get support. And the mqjms is just a log file. Maybe we could just git ignore them ?

aldettinger commented on message-bridge/src/main/java/org/acme/message/bridge/Producers.java:
```diff
@@ -0,0 +1,83 @@
+     * @return    a new JmsComponent instance
+     */
+    @Identifier("ibmmq")
+    JmsComponent ibmmq(@Identifier("ibmConnectionFactory") ConnectionFactory cf, JtaTransactionManager tm) {
```
From local debug, the `JtaTransactionManager tm` reference is not the same between `ibmmq` and `amq`:

```
ibm tm: org.springframework.transaction.jta.JtaTransactionManager@261db982
amq tm: org.springframework.transaction.jta.JtaTransactionManager@3a90c13c
```

@zhfeng @jamesnetherton Does one recall if this an issue ? Like we would need a singleton transaction manager to have both XAResource part of the same XA transaction ? Or maybe it's fine like this ?

zhfeng replied:
I think it should use the same `JtaTansactionManager`.

aldettinger replied:
ok, so let's mark the jta transaction manager as singleton in that case.

aldettinger commented:
Looks Narayana is doing the actual 2 Phase Commit under the hood. And as such, the transaction logs are stored in the `ObjectStore` folder. At least, it looks that way according to logs below:

```
$ inotifywait -m -r .

./ObjectStore/ShadowNoFileLockStore/defaultStore/StateManager/BasicAction/TwoPhaseCoordinator/AtomicAction/ CREATE 0_ffff7f000001_b517_65268954_1102
./ObjectStore/ShadowNoFileLockStore/defaultStore/StateManager/BasicAction/TwoPhaseCoordinator/AtomicAction/ OPEN 0_ffff7f000001_b517_65268954_1102
./ObjectStore/ShadowNoFileLockStore/defaultStore/StateManager/BasicAction/TwoPhaseCoordinator/AtomicAction/ MODIFY 0_ffff7f000001_b517_65268954_1102
./ObjectStore/ShadowNoFileLockStore/defaultStore/StateManager/BasicAction/TwoPhaseCoordinator/AtomicAction/ CLOSE_WRITE,CLOSE 0_ffff7f000001_b517_65268954_1102
./ObjectStore/ShadowNoFileLockStore/defaultStore/StateManager/BasicAction/TwoPhaseCoordinator/AtomicAction/ DELETE 0_ffff7f000001_b517_65268954_1102
```

Is it worth to explicitly add a note in the `README.adoc` file to explain where the transaction logs are being persisted ?
Folks may find this useful to know where the state is stored. For instance, when deploying to kubernetes, there could be a need to configure volume/pvc not to lose the state.

Which by the way, introduce another question. Would it be useful in the future to maybe deploy this example to kubernetes ala [file-bindy-ftp](https://github.com/apache/camel-quarkus-examples/tree/main/file-bindy-ftp/src/main/kubernetes)... Maybe not priority, let see this in another pr.

aldettinger commented:
By the way, this is really a good starter @avano. We may well find some users that would find this example extremely helpful :clap:

zhfeng commented:
Good catch @aldettinger - For using transaction in clould environment, it would be good to use `jdbc-object-store` and will persist the transaction logs into the database. Please refer to https://quarkus.io/guides/transaction#jdbcstore

aldettinger commented:
> Good catch @aldettinger - For using transaction in clould environment, it would be good to use `jdbc-object-store` and will persist the transaction logs into the database. Please refer to https://quarkus.io/guides/transaction#jdbcstore

Apparently from the quoted doc, it's possible to use persistent volume or database.
So maybe, as a first step we could add a bit of doc in README ala:

```
Under the hood, the distributed transaction is performed through XA[internet link to xa].
As such, some transaction logs need to be persisted somewhere.
In this example, the transaction logs are persisted in the folder named OjbectStore in the current directory.
While deploying to kubernetes, one would need to use a persistent volume claim or to store the logs in a database[link to jdbc-object-store]
```

zhfeng commented:
The object store directory can be config by setting `quarkus.transaction-manager.object-store.directory`.

jamesnetherton commented:
Please can we add the `native` profile into `pom.xml` (I assume all of this stuff has native support?).

Also, as we're using Docker, can we add a profile to skip tests on Windows. Like [this](https://github.com/apache/camel-quarkus-examples/blob/main/cluster-leader-election/pom.xml#L340-L350).

jamesnetherton commented on message-bridge/pom.xml:
```diff
@@ -0,0 +1,274 @@
+    <description>Camel Quarkus Example :: Configure XA Transactions and connection pooling</description>
+
+    <properties>
+        <quarkus.platform.version>3.4.1</quarkus.platform.version>
```
```suggestion
        <quarkus.platform.version>3.5.0.CR1</quarkus.platform.version>
```

avano commented:
I updated the example with:
- added crash recovery scenario
- added openshift deployment steps
- switched tests to integration tests with testcontainers
- added ^ to readme

zhfeng commented on message-bridge/README.adoc:
```diff
@@ -0,0 +1,206 @@
+To trigger a JVM crash during the transaction, send a message containing the word `crash`.
+
+----
+curl -X POST -H "Content-Type: text/plain" http://localhost:8080/message -d 'Hello rollback'
```
I assume here should be `Hello crash`?

avano replied:
thanks, copy-paste error

aldettinger commented on message-bridge/README.adoc:
```diff
@@ -0,0 +1,206 @@
+
+This example comes with a couple of integration tests that verify the correct functionality. Before running the tests examine the IBM MQ license with:
+
+----
```
This one seems not to work, locally it outputs:

```
...
2023-10-16T13:49:24.309Z License not accepted
```

Is it really needed to run this docker container actually.
Indeed, it looks like the `mvn clean verify ...` command runs well without the docker command line executed.

avano replied:
technically it is not needed, but to run the ibm mq docker image, you need to accept the license. This only provides a way how to view the license in case you want to read it beforehand.

running the docker container with `-e LICENSE=view` only prints the license and ends with `License not accepted` as you mentioned.

this command comes from the container itself, e.g. if you run just `docker run icr.io/ibm-messaging/mq:9.3.2.1-r1` it ends with:

```
2023-10-17T06:14:37.166Z Error: Set environment variable LICENSE=accept to indicate acceptance of license terms and conditions.
2023-10-17T06:14:37.166Z License agreements and information can be viewed by setting the environment variable LICENSE=view.  You can also set the LANG environment variable to view the license in a different language.
2023-10-17T06:14:37.166Z Error checking license acceptance: Set environment variable LICENSE=accept to indicate acceptance of license terms and conditions
```

to run the tests, you need to use `-Dibm.mq.container.license=accept` and this value is passed to the docker image as the LICENSE env var [here](https://github.com/avano/camel-quarkus-examples/blob/ibmmq-amq/message-bridge/src/test/java/org/acme/message/bridge/resource/IBMMQTestResource.java#L40), otherwise if you don't accept the tests are [skipped](https://github.com/avano/camel-quarkus-examples/blob/ibmmq-amq/message-bridge/src/test/java/org/acme/message/bridge/MessageBridgeITCase.java#L43)

aldettinger replied:
Many thanks for sharing, it's understood now :+1: 

Thinking a bit more about the value of integration tests in examples. They actually prevent regressions, as we run the itests at each release when we [upgrade and tag examples](https://camel.apache.org/camel-quarkus/3.4.x/contributor-guide/release-guide.html#_upgrade_and_tag_examples). So, the rationale would be that it's not really up to the example user to read the license but more the developer that commit the code. We have kind of similar mechanism like [that](https://github.com/apache/camel-quarkus/blob/main/integration-test-groups/jdbc/mssql/src/test/resources/container-license-acceptance.txt).

In this case, shouldn't we remove the test section from the README ? And then set the `LICENSE=accept` environment variable in the test config to show that the developers accept the license in the context of itests ?

aldettinger commented on message-bridge/README.adoc:
```diff
@@ -0,0 +1,206 @@
+----
+curl -X POST -H "Content-Type: text/plain" http://$(oc get route camel-quarkus-examples-message-bridge -o jsonpath='{.spec.host}')/message -d 'Hello world'
+----
+
```
Would be good to have a cleanup section, locally this worked for me:

```
oc delete all --selector app=activemq-artemis-broker -o name
oc delete all --selector app=mq -o name
oc delete all --selector app.kubernetes.io/name=camel-quarkus-examples-message-bridge
oc delete pvc message-bridge
```

aldettinger replied:
Apart from that, the transaction recovery has also worked on a sandbox oc instance :clap:

aldettinger commented on message-bridge/README.adoc:
```diff
@@ -0,0 +1,206 @@
+To trigger a JVM crash during the transaction, send a message containing the word `crash`.
+
+----
+curl -X POST -H "Content-Type: text/plain" http://localhost:8080/message -d 'Hello crash'
```
Looks it doesn't work when re-executing the `mvn clean compile quarkus:dev` command.
We need not to run clean, only `mvn compile quarkus:dev`.

aldettinger commented on message-bridge/README.adoc:
```diff
@@ -0,0 +1,206 @@
+<truncated exception from the simulated rollback>
+----
+
+Because the transaction was rolled back, the message wasn't successfully processed and it is returned to the IBMMQ queue and subsequently redelivered. In this instance, the delivery is successful and logged accordingly:
```
nitpick, would rephrase `it is returned` into `it stays in the input broker`

aldettinger commented on message-bridge/src/main/java/org/acme/message/bridge/Producers.java:
```diff
@@ -0,0 +1,99 @@
+     * <p/>
+     * Since there is no IBM MQ extension for quarkus, we need to create the connection factory manually
+     *
+     * @param  wrapper wrapper that is used to add pooling capabilities to the connection facory
```
nitpick typo: `factory`

zhfeng commented:
Great work @avano - I'm going to merge it. @jamesnetherton @aldettinger do you have anything to add?

aldettinger reviewed (approved):
Big congrats @avano, really nice example :clap:  :clap:  :clap:

Time permitting, next steps could be to activate the native profile when supported. And maybe also enabling kubernetes deployment by refactoring like [other examples](https://github.com/apache/camel-quarkus-examples/tree/camel-quarkus-main/file-bindy-ftp/src/main/kubernetes).
